File: src/main/java/no/nav/bidrag/beregn/felles/SjablonUtil.java
package no.nav.bidrag.beregn.felles;

import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;
import no.nav.bidrag.beregn.felles.bo.Sjablon;
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold;
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel;
import no.nav.bidrag.beregn.felles.bo.SjablonSingelNokkel;
import no.nav.bidrag.beregn.felles.bo.SjablonSingelNokkelSingelInnhold;
import no.nav.bidrag.beregn.felles.bo.TrinnvisSkattesats;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNokkelNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;

public class SjablonUtil {

  // Henter verdier fra sjablonene Barnetilsyn (N:1, eksakt match) og Bidragsevne (1:N, eksakt match)
  public static BigDecimal hentSjablonverdi(List<Sjablon> sjablonListe, SjablonNavn sjablonNavn, List<SjablonNokkel> sjablonNokkelListe,
      SjablonInnholdNavn sjablonInnholdNavn) {
    var filtrertSjablonListe = filtrerSjablonNokkelListePaaSjablonNokkel(filtrerPaaSjablonNavn(sjablonListe, sjablonNavn.getNavn()),
        sjablonNokkelListe);
    var sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe);
    return hentSjablonInnholdVerdiEksakt(sjablonInnholdListe, sjablonInnholdNavn);
  }

  // Henter verdier fra sjablonene Forbruksutgifter, MaksFradrag og MaksTilsyn (1:1, intervall)
  public static BigDecimal hentSjablonverdi(List<Sjablon> sjablonListe, SjablonNavn sjablonNavn, Integer sjablonNokkelVerdi) {
    var filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe, sjablonNavn.getNavn());
    var sortertSjablonSingelNokkelSingelInnholdListe = mapTilSingelListeNokkelInnholdSortert(filtrertSjablonListe);
    return hentSjablonInnholdVerdiIntervall(sortertSjablonSingelNokkelSingelInnholdListe, sjablonNokkelVerdi);
  }

  // Henter verdier fra sjablon Samværsfradrag (N:N, eksakt match + intervall)
  public static BigDecimal hentSjablonverdi(List<Sjablon> sjablonListe, SjablonNavn sjablonNavn, List<SjablonNokkel> sjablonNokkelListe,
      SjablonNokkelNavn sjablonNokkelNavn, Integer sjablonNokkelVerdi, SjablonInnholdNavn sjablonInnholdNavn) {
    var filtrertSjablonListe = filtrerSjablonNokkelListePaaSjablonNokkel(filtrerPaaSjablonNavn(sjablonListe, sjablonNavn.getNavn()),
        sjablonNokkelListe);
    var sortertSjablonSingelNokkelListe = mapTilSingelListeNokkelSortert(filtrertSjablonListe, sjablonNokkelNavn);
    var sjablonInnholdListe = finnSjablonInnholdVerdiListeIntervall(sortertSjablonSingelNokkelListe, sjablonNokkelVerdi);
    return hentSjablonInnholdVerdiEksakt(sjablonInnholdListe, sjablonInnholdNavn);
  }

  // Henter verdier fra sjablon Sjablontall (1:1, eksakt match)
  public static BigDecimal hentSjablonverdi(List<Sjablon> sjablonListe, SjablonTallNavn sjablonTallNavn) {
    var filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe, sjablonTallNavn.getNavn());
    var sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe);
    return hentSjablonInnholdVerdiEksakt(sjablonInnholdListe, SjablonInnholdNavn.SJABLON_VERDI);
  }

  // Henter liste med verdier fra sjablon TrinnvisSkattesats (0:N, hent alle)
  public static List<TrinnvisSkattesats> hentTrinnvisSkattesats(List<Sjablon> sjablonListe, SjablonNavn sjablonNavn) {
    var filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe, sjablonNavn.getNavn());
    var sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe);
    var inntektGrenseListe = finnSjablonInnholdVerdiListe(sjablonInnholdListe, SjablonInnholdNavn.INNTEKTSGRENSE_BELOP);
    var satsListe = finnSjablonInnholdVerdiListe(sjablonInnholdListe, SjablonInnholdNavn.SKATTESATS_PROSENT);

    var indeks = 0;
    var trinnvisSkattesatsListe = new ArrayList<TrinnvisSkattesats>();
    while (indeks < inntektGrenseListe.size()) {
      trinnvisSkattesatsListe.add(new TrinnvisSkattesats(inntektGrenseListe.get(indeks), satsListe.get(indeks)));
      indeks = indeks + 1;
    }

    return trinnvisSkattesatsListe.stream().sorted(comparing(TrinnvisSkattesats::getInntektGrense)).collect(toList());
  }

  // Filtrerer sjablonListe på sjablonNavn og returnerer ny liste.
  // Brukes av alle typer sjabloner.
  private static List<Sjablon> filtrerPaaSjablonNavn(List<Sjablon> sjablonListe, String sjablonNavn) {
    return sjablonListe
        .stream()
        .filter(sjablon -> sjablon.getNavn().equals(sjablonNavn))
        .collect(toList());
  }

  // Filtrerer sjablonListe på sjablonNokkelListe og returnerer en ny liste.
  // Brukes av sjabloner som har eksakt match på nøkkel (Barnetilsyn, Bidragsevne, Samværsfradrag).
  private static List<Sjablon> filtrerSjablonNokkelListePaaSjablonNokkel(List<Sjablon> sjablonListe, List<SjablonNokkel> sjablonNokkelListe) {
    var sjablonStream = sjablonListe.stream();
    for (SjablonNokkel sjablonNokkel : sjablonNokkelListe) {
      sjablonStream = filtrerPaaSjablonNokkel(sjablonStream, sjablonNokkel);
    }
    return sjablonStream.collect(toList());
  }

  // Filtrerer sjablonStream på sjablonNokkelInput og returnerer en ny stream.
  // Intern bruk.
  private static Stream<Sjablon> filtrerPaaSjablonNokkel(Stream<Sjablon> sjablonStream, SjablonNokkel sjablonNokkelInput) {
    return sjablonStream
        .filter(sjablon -> sjablon.getNokkelListe()
            .stream()
            .anyMatch(sjablonNokkel -> (sjablonNokkel.getNavn().equals(sjablonNokkelInput.getNavn())) &&
                (sjablonNokkel.getVerdi().equals(sjablonNokkelInput.getVerdi()))));
  }

  // Tar inn en sjablonListe og returnerer en sjablonInnholdListe.
  // Brukes av Bidragsevne, Sjablontall, TrinnvisSkattesats.
  private static List<SjablonInnhold> mapSjablonListeTilSjablonInnholdListe(List<Sjablon> sjablonListe) {
    return sjablonListe.stream().map(Sjablon::getInnholdListe).flatMap(Collection::stream).collect(toList());
  }

  // Tar inn filtrertSjablonListe og mapper denne om til en liste med singel nøkkelverdi og singel innholdverdi (1:1). Returnerer en ny liste sortert
  // på nøkkelverdi.
  // Brukes av sjabloner som har ett nøkkelobjekt og ett innholdobjekt (Forbruksutgifter, MaxFradrag, MaxTilsyn).
  private static List<SjablonSingelNokkelSingelInnhold> mapTilSingelListeNokkelInnholdSortert(List<Sjablon> filtrertSjablonListe) {
    return filtrertSjablonListe
        .stream()
        .map(sjablon -> new SjablonSingelNokkelSingelInnhold(sjablon.getNavn(),
            sjablon.getNokkelListe()
                .stream()
                .map(SjablonNokkel::getVerdi)
                .findFirst()
                .orElse(" "),
            sjablon.getInnholdListe()
                .stream()
                .map(SjablonInnhold::getVerdi)
                .findFirst()
                .orElse(BigDecimal.ZERO)))
        .sorted(comparing(sjablonSingelNokkelSingelInnhold -> Integer.valueOf(sjablonSingelNokkelSingelInnhold.getNokkelVerdi())))
        .collect(toList());
  }

  // Tar inn filtrertSjablonListe og mapper denne om til en liste med singel nøkkelverdi og liste med innholdverdier (1:N). Returnerer en ny liste
  // sortert på nøkkelverdi.
  // Brukes av sjabloner som har ett nøkkelobjekt med eksakt match og flere innholdobjekter (Samværsfradrag).
  private static List<SjablonSingelNokkel> mapTilSingelListeNokkelSortert(List<Sjablon> filtrertSjablonListe, SjablonNokkelNavn sjablonNokkelNavn) {
    return filtrertSjablonListe
        .stream()
        .map(sjablon -> new SjablonSingelNokkel(sjablon.getNavn(),
            sjablon.getNokkelListe()
                .stream()
                .filter(sjablonNokkel -> sjablonNokkel.getNavn().equals(sjablonNokkelNavn.getNavn()))
                .map(SjablonNokkel::getVerdi)
                .findFirst()
                .orElse(" "),
            sjablon.getInnholdListe()))
        .sorted(comparing(sjablonSingelNokkel -> Integer.valueOf(sjablonSingelNokkel.getVerdi())))
        .collect(toList());
  }

  // Filtrerer sjablonInnholdListe på sjablonInnholdNavn (eksakt match) og returnerer matchende verdi (0d hvis sjablonInnholdNavn mot formodning ikke
  // finnes).
  // Brukes av sjabloner som skal hente eksakt verdi (Barnetilsyn, Bidragsevne, Sjablontall, Samværsfradrag).
  private static BigDecimal hentSjablonInnholdVerdiEksakt(List<SjablonInnhold> sjablonInnholdListe, SjablonInnholdNavn sjablonInnholdNavn) {
    return sjablonInnholdListe
        .stream()
        .filter(sjablonInnhold -> sjablonInnhold.getNavn().equals(sjablonInnholdNavn.getNavn()))
        .map(SjablonInnhold::getVerdi)
        .findFirst()
        .orElse(BigDecimal.ZERO);
  }

  // Filtrerer sortertSjablonSingelNokkelSingelInnholdListe på nøkkel-verdi >= sjablonNokkel og returnerer en singel verdi (0d hvis det mot formodning
  // ikke finnes noen verdi).
  // Brukes av 1:1 sjabloner som henter verdi basert på intervall (Forbruksutgifter, MaxFradrag, MaxTilsyn).
  private static BigDecimal hentSjablonInnholdVerdiIntervall(List<SjablonSingelNokkelSingelInnhold> sortertSjablonSingelNokkelSingelInnholdListe,
      Integer sjablonNokkelVerdi) {
    return sortertSjablonSingelNokkelSingelInnholdListe
        .stream()
        .filter(sortertSjablonSingelNokkelSingelInnhold ->
            Integer.valueOf(sortertSjablonSingelNokkelSingelInnhold.getNokkelVerdi()).compareTo(sjablonNokkelVerdi) >= 0)
        .findFirst()
        .map(SjablonSingelNokkelSingelInnhold::getInnholdVerdi)
        .orElse(BigDecimal.ZERO);
  }

  // Filtrerer sortertSjablonSingelNokkelListe på nøkkel-verdi >= sjablonNokkel og returnerer en liste av typen SjablonInnholdNy (tom liste hvis det
  // mot formodning ikke finnes noen forekomster).
  // Brukes av sjabloner som har flere innholdobjekter og som henter verdi(er) basert på intervall (Samværsfradrag).
  private static List<SjablonInnhold> finnSjablonInnholdVerdiListeIntervall(List<SjablonSingelNokkel> sortertSjablonSingelNokkelListe,
      Integer sjablonNokkelVerdi) {
    return sortertSjablonSingelNokkelListe
        .stream()
        .filter(sjablon -> Integer.valueOf(sjablon.getVerdi()).compareTo(sjablonNokkelVerdi) >= 0)
        .findFirst()
        .map(SjablonSingelNokkel::getInnholdListe)
        .orElse(emptyList());
  }


  // Filtrerer sjablonInnholdListe på sjablonInnholdNavn og returnerer en liste over alle matchende verdier.
  // Brukes av sjabloner som skal returnere en liste med innholdverdier (TrinnvisSkattesats).
  private static List<BigDecimal> finnSjablonInnholdVerdiListe(List<SjablonInnhold> sjablonInnholdListe, SjablonInnholdNavn sjablonInnholdNavn) {
    return sjablonInnholdListe
        .stream()
        .filter(sjablonInnhold -> sjablonInnhold.getNavn().equals(sjablonInnholdNavn.getNavn()))
        .map(SjablonInnhold::getVerdi)
        .collect(toList());
  }
}

File: bidrag-beregn-core/src/main/kotlin/no/nav/bidrag/beregn/felles/util/SjablonUtil.kt
package no.nav.bidrag.beregn.felles.util

import no.nav.bidrag.beregn.felles.bo.Sjablon
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel
import no.nav.bidrag.beregn.felles.bo.SjablonSingelNokkel
import no.nav.bidrag.beregn.felles.bo.SjablonSingelNokkelSingelInnhold
import no.nav.bidrag.beregn.felles.bo.TrinnvisSkattesats
import no.nav.bidrag.domene.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNøkkelNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonTallNavn
import java.math.BigDecimal
import java.util.Comparator.comparing
import java.util.stream.Stream

object SjablonUtil {
    // Henter verdier fra sjablonene Barnetilsyn (N:1, eksakt match) og Bidragsevne (1:N, eksakt match)
    @JvmStatic
    fun hentSjablonverdi(
        sjablonListe: List<Sjablon>,
        sjablonNavn: SjablonNavn,
        sjablonNokkelListe: List<SjablonNokkel>,
        sjablonInnholdNavn: SjablonInnholdNavn,
    ): BigDecimal {
        val filtrertSjablonListe =
            filtrerSjablonNokkelListePaaSjablonNokkel(
                sjablonListe = filtrerPaaSjablonNavn(sjablonListe = sjablonListe, sjablonNavn = sjablonNavn.navn),
                sjablonNokkelListe = sjablonNokkelListe,
            )
        val sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe)

        return hentSjablonInnholdVerdiEksakt(sjablonInnholdListe = sjablonInnholdListe, sjablonInnholdNavn = sjablonInnholdNavn)
    }

    // Henter verdier fra sjablonene Forbruksutgifter, MaksFradrag og MaksTilsyn (1:1, intervall)
    @JvmStatic
    fun hentSjablonverdi(
        sjablonListe: List<Sjablon>,
        sjablonNavn: SjablonNavn,
        sjablonNokkelVerdi: Int,
    ): BigDecimal {
        val filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe = sjablonListe, sjablonNavn = sjablonNavn.navn)
        val sortertSjablonSingelNokkelSingelInnholdListe = mapTilSingelListeNokkelInnholdSortert(filtrertSjablonListe)

        return hentSjablonInnholdVerdiIntervall(
            sortertSjablonSingelNokkelSingelInnholdListe = sortertSjablonSingelNokkelSingelInnholdListe,
            sjablonNokkelVerdi = sjablonNokkelVerdi,
        )
    }

    // Henter verdier fra sjablon Samværsfradrag (N:N, eksakt match + intervall)
    @JvmStatic
    fun hentSjablonverdi(
        sjablonListe: List<Sjablon>,
        sjablonNavn: SjablonNavn,
        sjablonNokkelListe: List<SjablonNokkel>,
        sjablonNokkelNavn: SjablonNøkkelNavn,
        sjablonNokkelVerdi: Int,
        sjablonInnholdNavn: SjablonInnholdNavn,
    ): BigDecimal {
        val filtrertSjablonListe =
            filtrerSjablonNokkelListePaaSjablonNokkel(
                sjablonListe = filtrerPaaSjablonNavn(sjablonListe = sjablonListe, sjablonNavn = sjablonNavn.navn),
                sjablonNokkelListe = sjablonNokkelListe,
            )
        val sortertSjablonSingelNokkelListe =
            mapTilSingelListeNokkelSortert(
                filtrertSjablonListe = filtrertSjablonListe,
                sjablonNokkelNavn = sjablonNokkelNavn,
            )
        val sjablonInnholdListe =
            finnSjablonInnholdVerdiListeIntervall(
                sortertSjablonSingelNokkelListe = sortertSjablonSingelNokkelListe,
                sjablonNokkelVerdi = sjablonNokkelVerdi,
            )

        return hentSjablonInnholdVerdiEksakt(sjablonInnholdListe = sjablonInnholdListe, sjablonInnholdNavn = sjablonInnholdNavn)
    }

    // Henter verdier fra sjablon Sjablontall (1:1, eksakt match)
    @JvmStatic
    fun hentSjablonverdi(
        sjablonListe: List<Sjablon>,
        sjablonTallNavn: SjablonTallNavn,
    ): BigDecimal {
        val filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe = sjablonListe, sjablonNavn = sjablonTallNavn.navn)
        val sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe)
        return hentSjablonInnholdVerdiEksakt(
            sjablonInnholdListe = sjablonInnholdListe,
            sjablonInnholdNavn = SjablonInnholdNavn.SJABLON_VERDI,
        )
    }

    // Henter liste med verdier fra sjablon TrinnvisSkattesats (0:N, hent alle)
    @JvmStatic
    fun hentTrinnvisSkattesats(
        sjablonListe: List<Sjablon>,
        sjablonNavn: SjablonNavn,
    ): List<TrinnvisSkattesats> {
        val filtrertSjablonListe = filtrerPaaSjablonNavn(sjablonListe = sjablonListe, sjablonNavn = sjablonNavn.navn)
        val sjablonInnholdListe = mapSjablonListeTilSjablonInnholdListe(filtrertSjablonListe)
        val inntektGrenseListe =
            finnSjablonInnholdVerdiListe(
                sjablonInnholdListe = sjablonInnholdListe,
                sjablonInnholdNavn = SjablonInnholdNavn.INNTEKTSGRENSE_BELØP,
            )
        val satsListe =
            finnSjablonInnholdVerdiListe(
                sjablonInnholdListe = sjablonInnholdListe,
                sjablonInnholdNavn = SjablonInnholdNavn.SKATTESATS_PROSENT,
            )
        var indeks = 0
        val trinnvisSkattesatsListe = ArrayList<TrinnvisSkattesats>()
        while (indeks < inntektGrenseListe.size) {
            trinnvisSkattesatsListe.add(TrinnvisSkattesats(inntektGrense = inntektGrenseListe[indeks], sats = satsListe[indeks]))
            indeks += 1
        }

        return trinnvisSkattesatsListe.sortedWith(comparing(TrinnvisSkattesats::inntektGrense)).toList()
    }

    // Filtrerer sjablonListe på sjablonNavn og returnerer ny liste.
    // Brukes av alle typer sjabloner.
    private fun filtrerPaaSjablonNavn(
        sjablonListe: List<Sjablon>,
        sjablonNavn: String,
    ) = sjablonListe.filter { it.navn == sjablonNavn }.toList()

    // Filtrerer sjablonListe på sjablonNokkelListe og returnerer en ny liste.
    // Brukes av sjabloner som har eksakt match på nøkkel (Barnetilsyn, Bidragsevne, Samværsfradrag).
    private fun filtrerSjablonNokkelListePaaSjablonNokkel(
        sjablonListe: List<Sjablon>,
        sjablonNokkelListe: List<SjablonNokkel>,
    ): List<Sjablon> {
        var sjablonStream = sjablonListe.stream()
        sjablonNokkelListe.forEach {
            sjablonStream = filtrerPaaSjablonNokkel(sjablonStream, it)
        }
        return sjablonStream.toList()
    }

    // Filtrerer sjablonStream på sjablonNokkelInput og returnerer en ny stream.
    // Intern bruk.
    private fun filtrerPaaSjablonNokkel(
        sjablonStream: Stream<Sjablon>,
        sjablonNokkelInput: SjablonNokkel,
    ) = sjablonStream.filter {
        it.nokkelListe!!
            .any { sjablonNokkel ->
                sjablonNokkel.navn == sjablonNokkelInput.navn &&
                    sjablonNokkel.verdi == sjablonNokkelInput.verdi
            }
    }

    // Tar inn en sjablonListe og returnerer en sjablonInnholdListe.
    // Brukes av Bidragsevne, Sjablontall, TrinnvisSkattesats.
    private fun mapSjablonListeTilSjablonInnholdListe(sjablonListe: List<Sjablon>) = sjablonListe.flatMap { it.innholdListe }

    // Tar inn filtrertSjablonListe og mapper denne om til en liste med singel nøkkelverdi og singel innholdverdi (1:1). Returnerer en ny liste sortert
    // på nøkkelverdi.
    // Brukes av sjabloner som har ett nøkkelobjekt og ett innholdobjekt (Forbruksutgifter, MaxFradrag, MaxTilsyn).
    private fun mapTilSingelListeNokkelInnholdSortert(filtrertSjablonListe: List<Sjablon>) =
        filtrertSjablonListe
            .map {
                SjablonSingelNokkelSingelInnhold(
                    navn = it.navn,
                    nokkelVerdi = it.nokkelListe!!.firstOrNull()?.verdi ?: " ",
                    innholdVerdi = it.innholdListe.firstOrNull()?.verdi ?: BigDecimal.ZERO,
                )
            }
            .sortedBy { Integer.valueOf(it.nokkelVerdi) }

    // Tar inn filtrertSjablonListe og mapper denne om til en liste med singel nøkkelverdi og liste med innholdverdier (1:N). Returnerer en ny liste
    // sortert på nøkkelverdi.
    // Brukes av sjabloner som har ett nøkkelobjekt med eksakt match og flere innholdobjekter (Samværsfradrag).
    private fun mapTilSingelListeNokkelSortert(
        filtrertSjablonListe: List<Sjablon>,
        sjablonNokkelNavn: SjablonNøkkelNavn,
    ): List<SjablonSingelNokkel> =
        filtrertSjablonListe
            .map {
                SjablonSingelNokkel(
                    navn = it.navn,
                    verdi =
                        it.nokkelListe!!
                            .filter { sjablonNokkel ->
                                sjablonNokkel.navn == sjablonNokkelNavn.navn
                            }
                            .map(SjablonNokkel::verdi)
                            .firstOrNull() ?: " ",
                    innholdListe = it.innholdListe,
                )
            }
            .sortedBy { Integer.valueOf(it.verdi) }

    // Filtrerer sjablonInnholdListe på sjablonInnholdNavn (eksakt match) og returnerer matchende verdi (0d hvis sjablonInnholdNavn mot formodning ikke
    // finnes).
    // Brukes av sjabloner som skal hente eksakt verdi (Barnetilsyn, Bidragsevne, Sjablontall, Samværsfradrag).
    private fun hentSjablonInnholdVerdiEksakt(
        sjablonInnholdListe: List<SjablonInnhold>,
        sjablonInnholdNavn: SjablonInnholdNavn,
    ) = sjablonInnholdListe
        .filter { it.navn == sjablonInnholdNavn.navn }
        .map { it.verdi }
        .firstOrNull() ?: BigDecimal.ZERO

    // Filtrerer sortertSjablonSingelNokkelSingelInnholdListe på nøkkel-verdi >= sjablonNokkel og returnerer en singel verdi (0d hvis det mot formodning
    // ikke finnes noen verdi).
    // Brukes av 1:1 sjabloner som henter verdi basert på intervall (Forbruksutgifter, MaxFradrag, MaxTilsyn).

    private fun hentSjablonInnholdVerdiIntervall(
        sortertSjablonSingelNokkelSingelInnholdListe: List<SjablonSingelNokkelSingelInnhold>,
        sjablonNokkelVerdi: Int,
    ) = sortertSjablonSingelNokkelSingelInnholdListe
        .filter { it.nokkelVerdi.toInt() >= sjablonNokkelVerdi }
        .map { it.innholdVerdi }
        .firstOrNull() ?: BigDecimal.ZERO

    // Filtrerer sortertSjablonSingelNokkelListe på nøkkel-verdi >= sjablonNokkel og returnerer en liste av typen SjablonInnholdNy (tom liste hvis det
    // mot formodning ikke finnes noen forekomster).
    // Brukes av sjabloner som har flere innholdobjekter og som henter verdi(er) basert på intervall (Samværsfradrag).
    private fun finnSjablonInnholdVerdiListeIntervall(
        sortertSjablonSingelNokkelListe: List<SjablonSingelNokkel>,
        sjablonNokkelVerdi: Int,
    ) = sortertSjablonSingelNokkelListe
        .filter { it.verdi.toInt() >= sjablonNokkelVerdi }
        .map { it.innholdListe }
        .firstOrNull() ?: emptyList()

    // Filtrerer sjablonInnholdListe på sjablonInnholdNavn og returnerer en liste over alle matchende verdier.
    // Brukes av sjabloner som skal returnere en liste med innholdverdier (TrinnvisSkattesats).
    private fun finnSjablonInnholdVerdiListe(
        sjablonInnholdListe: List<SjablonInnhold>,
        sjablonInnholdNavn: SjablonInnholdNavn,
    ) = sjablonInnholdListe
        .filter { it.navn == sjablonInnholdNavn.navn }
        .map { it.verdi }
        .toList()
}
